crates/nargo_cli/src/cli/fs/common_reference_string.rs

```diff
@@ -0,0 +1,55 @@
+use std::{env, path::PathBuf};
+
+use acvm::{acir::circuit::Circuit, CommonReferenceString};
+
+use super::{create_named_dir, write_to_file};
+
+// TODO: pull this from backend.
+const BACKEND_IDENTIFIER: &str = "acvm-backend-barretenberg";
+const TRANSCRIPT_NAME: &str = "common-reference-string.bin";
+
+fn common_reference_string_location() -> PathBuf {
+    let cache_dir = match env::var("NARGO_BACKEND_CACHE_DIR") {
+        Ok(cache_dir) => PathBuf::from(cache_dir),
+        Err(_) => dirs::home_dir().unwrap().join(".nargo").join("backends"),
+    };
+    cache_dir.join(BACKEND_IDENTIFIER).join(TRANSCRIPT_NAME)
+}
+
+pub(crate) fn read_cached_common_reference_string() -> Vec<u8> {
+    let crs_path = common_reference_string_location();
+
+    // TODO: Implement checksum
+    match std::fs::read(crs_path) {
+        Ok(common_reference_string) => common_reference_string,
+        Err(_) => vec![],
+    }
+}
+
+pub(crate) fn update_common_reference_string<B: CommonReferenceString>(
+    backend: &B,
+    common_reference_string: &[u8],
+    circuit: &Circuit,
+) -> Result<Vec<u8>, B::Error> {
+    use tokio::runtime::Builder;
+
+    let runtime = Builder::new_current_thread().enable_all().build().unwrap();
+
+    // TODO: Implement retries
+    // If the read data is empty, we don't have a CRS and need to generate one
+    let fut = if common_reference_string.is_empty() {
+        backend.generate_common_reference_string(circuit)
+    } else {
+        backend.update_common_reference_string(common_reference_string.to_vec(), circuit)
+    };
```

Comment:
```suggestion
    let fut = backend.update_common_reference_string(common_reference_string.to_vec(), circuit);
```

This feels like something that the backend should handle. If we pass an empty vector in here do we not get a proper CRS out?

Reply:
Ah, no it won't because it'll just fail to deserialise.